### PR TITLE
Tag filtering in <Student.TracksList />

### PR DESCRIPTION
**Pull request by kntsoriano**
Closes https://github.com/exercism/v3-project-management/issues/49.
Closes https://github.com/exercism/v3-project-management/issues/48.

## Preview
Oct-05-2020 12-50-28

### CODE REVIEW

File: app/helpers/view_components/student/tracks_list.rb
Comment (kntsoriano) [inline app/helpers/view_components/student/tracks_list.rb:24]: Can you help me with this?

1. Is the structure okay?
2. Is the naming okay?

Comment (iHiD) [reply app/helpers/view_components/student/tracks_list.rb:24]: Yeah - this is totally fine for now. We'll probably change it once we've thought it through further, but this works for now 🙂

File: app/controllers/test/components/student/tracks_list_controller.rb
Comment (iHiD) [inline app/controllers/test/components/student/tracks_list_controller.rb]: Is this a pattern we're planning on using in general?

If so, I'd like to standardise it so we have a helper in JS land that serializes to this, and a helper in Rails land that deserializes. It feels a little too arbitary to me to have this LOC in multiple places without it being  formalised as a pattern.

That's not blocking this being merged, but if you agree, can you open an issue to do that when you merge please.

We should also consider what character to use - there's probably a better choice than a `,`, which might actually be part of the string in some places. Maybe a unicode character that we have confidence wouldn't be used in a string?

**Review by iHiD — APPROVED** [review]
This seems great to me. Let Erik have a review as well pls before merging.

File: app/controllers/test/components/student/tracks_list_controller.rb
Comment (SleeplessByte) [reply app/controllers/test/components/student/tracks_list_controller.rb]: Instead of using a `,`, use the HTML standard and push `tags` as an array: 

```
tags[]=foo&tags[]=bar

params[:tags]
# => ['foo', 'bar']
```

Comment (iHiD) [reply app/controllers/test/components/student/tracks_list_controller.rb]: @SleeplessByte Do you know of a widly used React (or JS or whatever) helper for this, which we should use?

Comment (SleeplessByte) [reply app/controllers/test/components/student/tracks_list_controller.rb]: For autocomplete? Something like this: https://github.com/alphagov/accessible-autocomplete, or maybe [this fork](https://bitzesty.com/2020/09/25/making-gov-uk-accessible-autocomplete-a-suitable-replacement-for-select-enhancement/?utm_content=buffer992c3&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer). Almost anything else, such as `selectize` is **not** accessible.

Comment (iHiD) [reply app/controllers/test/components/student/tracks_list_controller.rb]: Sorry, no, I just meant the serialization to the HTML array standard thing within React, before the data is passed to the server.

Comment (SleeplessByte) [reply app/controllers/test/components/student/tracks_list_controller.rb]: Ahhh. Got it. 

- If you want to push it to a URL: `URLSearchParams` (modern browsers, or [polyfill](https://github.com/ungap/url-search-params))
- If you want to push it as form data: `FormData` (built-in).

For URLs:  multiple values will have the same key, not an array key, so it will become `?tag=foo&tag=bar`, which is valid in browsers, dunno what Rails does, but this is correct behaviour.

But usually I just role my own. It's nothing more than `key[]={encodeUriComponent(value)}`

Comment (iHiD) [reply app/controllers/test/components/student/tracks_list_controller.rb]: @kntsoriano Shall we go with this instead to be more HTML compliant?

Comment (SleeplessByte) [reply app/controllers/test/components/student/tracks_list_controller.rb]: FWIW, comma separated works too, but yeah, then you have to deal with a comma never being a character inside a tag.

Comment (kntsoriano) [reply app/controllers/test/components/student/tracks_list_controller.rb]: Thanks! Applied the changes now.

**Comment by kntsoriano** [comment]
@iHiD I've updated this PR to include the solution to https://github.com/exercism/v3-project-management/issues/48. Can you take a look at it again if you have a chance? Thanks.

**Comment by iHiD** [comment]
Sweet. I'm happy with this. I'll let @ErikSchierboom review and merge if he's happy.

File: app/javascript/components/student/tracks-list/TagOption.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagOption.jsx:4]: Is there any chance we'll have another component render tag options? If so, there might be ID duplicates.

File: app/javascript/components/student/tracks-list/Header.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/Header.jsx]: Just a tiny suggestion:

```suggestion
  const isFiltering = query.criteria || (query.tags && query.tags.length > 0)
```

I feel that `> 0` is slightly easier to read than `!== 0`.

File: app/javascript/components/student/tracks-list/TagOptionList.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagOptionList.jsx]: What browsers are we supporting @iHiD? Are we supporting Internet Explorer?

If not, the [`includes`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes) function could possibly be used here.

File: app/javascript/components/student/tracks-list/TagOptionList.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagOptionList.jsx]: I didn't immediately understand what `value` represented. We do use it in several other places, so I can understand why we're using it. Just wanted to put out an alternative:

```suggestion
export function TagOptionList({ options, onSubmit, selectedTags, setSelectedTags }) {
```

File: app/javascript/components/student/tracks-list/TagsFilter.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagsFilter.jsx]: I didn't immediately understand what `value` represented. We do use it in several other places, so I can understand why we're using it. Just wanted to put out an alternative:

```suggestion
  const [selectedTags, setSelectedTags] = useState([])
```

or

```suggestion
  const [filterTags, setFilterTags] = useState([])
```

File: app/javascript/components/student/tracks-list/TagsFilter.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagsFilter.jsx]:
```suggestion
        onClick={() => setHidden(!hidden)}
```

File: app/javascript/components/student/tracks-list/TagsFilter.jsx
Comment (ErikSchierboom) [inline app/javascript/components/student/tracks-list/TagsFilter.jsx]:
```suggestion
            onClick={() => setHidden(true)}
```

**Review by ErikSchierboom — APPROVED** [review]
Nice!

File: app/javascript/utils/url-params.js
Comment (SleeplessByte) [inline app/javascript/utils/url-params.js]:
```suggestion
          return key + '=' + encodeURIComponent(value)
```

File: app/javascript/utils/url-params.js
Comment (SleeplessByte) [inline app/javascript/utils/url-params.js]:
```suggestion
              return key + '[]=' + encodeURIComponent(item)
```

File: test/javascript/utils/url-params.test.js
Comment (SleeplessByte) [inline test/javascript/utils/url-params.test.js]:
```suggestion
  const object = { tags: ['oop', 'static'], foo: 'bar' }
```

File: test/javascript/utils/url-params.test.js
Comment (SleeplessByte) [inline test/javascript/utils/url-params.test.js]:
```suggestion
  expect(new UrlParams(object).toString()).toEqual('tags[]=oop&tags[]=static&foo=bar')
```

File: app/javascript/utils/url-params.js
Comment (SleeplessByte) [inline app/javascript/utils/url-params.js:10]: You might want to invert this:

```typescript
 return Object.keys(this.object)
   .map((key) => {
     const value = this.object[key]

     // TODO: should check the spec if these should be _left out_ or should be added
     //             like this. If they should be left out, instead use a filter above the .map
     // &foo
     if (value === null || value === undefined || value === '') {
       return key
     }
     
     // &foo=value
     if (!Array.isArray(value)) {
       return `${key}=${encodeURIComponent(value)}`
     }
     
     // &foo[]=value1&foo[]=value2
     return value
       .map((item) => `${key}[]=${encodeURIComponent(item)}`)
       .join('&')
  })
  .join('&')
```

File: app/javascript/components/student/tracks-list/TagOptionList.jsx
Comment (SleeplessByte) [reply app/javascript/components/student/tracks-list/TagOptionList.jsx]: Transpilation of this can be configured in babel. I think it should be `includes` anyway.

File: app/javascript/components/student/tracks-list/TagOption.jsx
Comment (kntsoriano) [reply app/javascript/components/student/tracks-list/TagOption.jsx:4]: Good point. This is something to remember in the future.

File: app/javascript/components/student/tracks-list/Header.jsx
Comment (kntsoriano) [reply app/javascript/components/student/tracks-list/Header.jsx]: Sounds good!

**Comment by kntsoriano** [comment]
Thanks everyone! Merging this in now.

**Pull request merged**
